ocaml/xapi/repository.ml

```diff
@@ -457,41 +455,62 @@ let apply ~__context ~host =
       in
       try ignore (Helpers.call_script !Xapi_globs.yum_cmd params)
       with e ->
-        let ref = Ref.string_of host in
-        error "Failed to apply updates on host ref='%s': %s" ref
+        let host' = Ref.string_of host in
+        error "Failed to apply updates on host ref='%s': %s" host'
           (ExnHelper.string_of_exn e) ;
-        raise Api_errors.(Server_error (apply_updates_failed, [ref])))
+        raise Api_errors.(Server_error (apply_updates_failed, [host'])))
 
-let restart_device_models ~__context host =
-  (* Restart device models of all running HVM VMs on the host by doing
-   * local migrations. *)
+let do_with_device_models ~__context ~host ~action_label f =
+  (* Call f with device models of all running HVM VMs on the host *)
   Db.Host.get_resident_VMs ~__context ~self:host
   |> List.map (fun self -> (self, Db.VM.get_record ~__context ~self))
   |> List.filter (fun (_, record) -> not record.API.vM_is_control_domain)
-  |> List.filter_map (fun (ref, record) ->
-         match
-           ( record.API.vM_power_state
-           , Helpers.has_qemu_currently ~__context ~self:ref )
-         with
-         | `Running, true ->
-             Helpers.call_api_functions ~__context (fun rpc session_id ->
-                 Client.Client.VM.pool_migrate rpc session_id ref host
-                   [("live", "true")]) ;
-             None
-         | `Paused, true ->
-             error "VM 'ref=%s' is paused, can't restart its device models"
-               (Ref.string_of ref) ;
-             Some ref
-         | _ ->
-             (* No device models are running for this VM *)
-             None)
+  |> List.filter_map f
   |> function
   | [] ->
       ()
   | _ :: _ ->
-      let msg = "Can't restart device models for some VMs" in
+      let msg =
+        Printf.sprintf "Can't %s device models for some VMs" action_label
+      in
       raise Api_errors.(Server_error (internal_error, [msg]))
```

Comment:
It looks like this would occur (only) for paused VMs. It's not an internal error then, as anyone can pause VMs. It should be an explicit API error that the client can handle.